pkg/controllers/provisioning/v1alpha1/reallocation/collect.go

```diff
@@ -0,0 +1,121 @@
+package reallocation
+
+import (
+	"context"
+	"fmt"
+	"github.com/awslabs/karpenter/pkg/apis/provisioning/v1alpha1"
+	f "github.com/awslabs/karpenter/pkg/utils/functional"
+	utilsnode "github.com/awslabs/karpenter/pkg/utils/node"
+	"github.com/awslabs/karpenter/pkg/utils/ptr"
+	"go.uber.org/zap"
+	v1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"time"
+)
+
+type Collector struct {
+	kubeClient client.Client
+}
+
+func (c *Collector) ManageUnderutilization(ctx context.Context, provisioner *v1alpha1.Provisioner) error {
+	// 1. Set TTL on TTLable Nodes
+	if err := c.markUnderutilized(ctx, provisioner); err != nil {
+		return fmt.Errorf("adding ttl and underutilized label, %w", err)
+	}
+
+	// 2. Remove TTL from Utilized Nodes
+	if err := c.clearUnderutilized(ctx, provisioner); err != nil {
+		return fmt.Errorf("removing ttl from node, %w", err)
+	}
+	return nil
+}
+
+// markUnderutilized adds a TTL to underutilized nodes
+func (c *Collector) markUnderutilized(ctx context.Context, provisioner *v1alpha1.Provisioner) error {
+	ttlable := []*v1.Node{}
+
+	// 1. Get all provisioner labeled nodes
+	nodes, err := c.getNodes(ctx, provisioner, map[string]string{})
+	if err != nil {
+		return err
+	}
+
+	// 2. Get nodes and pods for each node
+	for _, node := range nodes {
+		pods, err := c.getPodsOnNode(ctx, node.Name)
+		if err != nil {
+			zap.S().Debugf("Unable to get pods for node %s, %s", node.Name, err.Error())
+			continue
```

Comment:
We probably don't want to continue here. If we fail to call the API Server (overloaded, or some other reason), we probably don't want to keep hammering it. Best to let the reconcile loop backoff/retry logic kick in and just return an error here.

Reply:
Agreed - my comment about sometimes continuing was specifically (only) for certain EC2 API calls (especially Terminate)